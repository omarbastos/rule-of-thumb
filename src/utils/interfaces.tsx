import firestore from 'firebase/firestore'

export interface Poll {
  name: string
  description: string
  category: string
  picture: string
  lastUpdated: firestore.Timestamp
  votes: Votes
  uid: string
}

export interface Votes {
  positive: number
  negative: number
}

export type CurrentVote = string | null
