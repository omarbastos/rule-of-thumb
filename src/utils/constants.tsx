export const CURRENT_VOTES = {
  POSITIVE: 'POSITIVE',
  NEGATIVE: 'NEGATIVE'
}
export const VIEWS = {
  LIST: 'List',
  GRID: 'Grid'
}
