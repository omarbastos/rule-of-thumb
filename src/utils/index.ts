import { Votes } from './interfaces'

export function classNames(...classes: unknown[]): string {
  return classes.filter(Boolean).join(' ')
}

export function getPercentage({ positive, negative }: Votes): {
  positive: number
  negative: number
} {
  const total = positive + negative
  if (total === 0) return { positive: 50, negative: 50 }
  const positivePercentage = +((positive / total) * 100).toFixed(2)
  const negativePercentage = +((negative / total) * 100).toFixed(2)
  return { positive: positivePercentage, negative: negativePercentage }
}
export const getImagePath = (image: string) => {
  return `../assets/img/${image}`
}
