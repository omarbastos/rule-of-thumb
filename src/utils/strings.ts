interface ReplaceInterface {
  value: string
  oldValue: string
  newValue: string
}

interface STGInterface<T> {
  replace: {
    [K: string]: T
  }
}

type StringsType = typeof strings

const strings = {
  speak_out: 'Speak out. Be heard.',
  be_counted: 'Be counted',
  banner_top_description: `Rule of Thumb is a crowd sourced court of public opinion where anyone
  and everyone can speak out and speak freely. It’s easy: You share your
  opinion, we analyze and put the data in a public report.`,
  banner_bottom_title: ' Is there anyone else you would want us to add?',
  submit_a_name: 'Submit a name',
  terms_and_conditions: 'Terms and Conditions',
  privacy_policy: 'Privacy Policy',
  contact_us: 'Contact Us',
  follow_us: 'Follow Us',
  rule_of_thumb: 'Rule of thumb.',
  past_trials: 'Past Trials',
  how_it_works: 'How It Works',
  login_signup: 'Login / Sign Up',
  vote_now: 'Vote Now',
  thank_you: 'Thank you for your vote!',
  vote_again: 'Vote Again',
  previous_rulings: 'Previous Rulings',
  what_is_your_opinion: "What's your opinion on",
  pope_francis: 'Pope Francis',
  hero_description: `He’s talking tough on clergy sexual abuse, or is he just another
  pervert protector? (thumbs down) or a true pedophile punishing
  pontiff? (thumbs up)`,
  more_information: 'More information',
  what_is_your_veredict: 'What’s Your Veredict?',
  closing_in: 'closing in',
  days: 'days'
}

const replaceValue = ({
  value,
  oldValue,
  newValue
}: ReplaceInterface): string => {
  if (value.replaceAll) return value.replaceAll(oldValue, newValue)
  return value
}

const applyOptions = <T extends string, K>(
  value: T,
  options: STGInterface<K>
): string => {
  const { replace } = options
  let vTemp: string = value
  if (replace) {
    Object.keys(replace).forEach((key) => {
      vTemp = replaceValue({
        value: vTemp,
        oldValue: key,
        newValue: `${replace[key]}`
      })
    })
  }
  return vTemp
}

const stg = <T extends keyof StringsType, K>(
  key: T,
  options?: STGInterface<K>
): string => {
  if (options) {
    return applyOptions(strings[key], options)
  }
  return strings[key]
}

export default stg
