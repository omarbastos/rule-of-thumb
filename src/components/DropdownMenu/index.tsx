import React, { Dispatch, SetStateAction, useRef, useState } from 'react'
import { VIEWS } from 'utils/constants'
import useOnClickOutside from 'utils/hooks/useOnClickOutside'

interface Props {
  selected: string
  setSelected: Dispatch<SetStateAction<string>>
}

const DropdownMenu = ({ setSelected, selected }: Props) => {
  const [isVisible, setIsVisible] = useState<boolean>(false)
  const itemsMenu = [
    {
      id: '1',
      title: VIEWS.LIST
    },
    { id: '2', title: VIEWS.GRID }
  ]
  const handleSelected = (title: string) => {
    setSelected(title)
    setIsVisible(!isVisible)
  }
  const ref = useRef(null)
  useOnClickOutside(ref, () => setIsVisible(false))

  return (
    <div className="relative flex-none" ref={ref}>
      <button
        aria-label="View"
        onClick={() => setIsVisible(!isVisible)}
        className="w-40 bg-white text-sm py-2 px-4 whitespace-nowrap text-black border-solid border-2 border-black font-normal flex justify-center items-center"
        type="button"
        data-testid="button-dropdow-menu"
      >
        <span className="flex-auto"> {selected}</span>
        <svg
          className=" flex-none justify-self-end"
          width={11}
          height={7}
          viewBox="0 0 11 7"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M5.25 7L6.11959e-07 -9.17939e-07L10.5 0L5.25 7Z"
            fill="#333333"
          />
        </svg>
      </button>
      {isVisible && (
        <div className="origin-top-left absolute left-0 w-full drop-shadow-lg overflow-y-auto max-h-32 z-10 divide-y-2">
          <ul className="cursor-pointer font-normal    text-sm bg-white  whitespace-nowrap text-black ">
            {itemsMenu.map((item) => {
              return (
                <li
                  data-testid={`dropdown-option-${item.id}`}
                  className="border-solid border-b-2 border-l-2 border-r-2 border-black py-2 px-4 flex justify-center items-center"
                  onClick={() => handleSelected(item.title)}
                  key={item.id}
                >
                  {item.title}
                </li>
              )
            })}
          </ul>
        </div>
      )}
    </div>
  )
}

export default DropdownMenu
