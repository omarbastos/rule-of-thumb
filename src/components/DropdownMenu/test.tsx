import { fireEvent, render } from '@testing-library/react'
import DropdownMenu from '.'
import { VIEWS } from '../../utils/constants'

describe('<DropdownMenu />', () => {
  it('should render DropdownMenu as default', () => {
    const mockSetSelected = jest.fn()
    const { container } = render(
      <DropdownMenu selected={VIEWS.GRID} setSelected={mockSetSelected} />
    )

    expect(container).toMatchSnapshot()
  })

  it('should called mockSetSelected when select an option', () => {
    const mockSetSelected = jest.fn()
    const { getByTestId } = render(
      <DropdownMenu selected={VIEWS.GRID} setSelected={mockSetSelected} />
    )
    fireEvent.click(getByTestId('button-dropdow-menu'))
    fireEvent.click(getByTestId('dropdown-option-1'))
    expect(mockSetSelected).toBeCalled()
  })
})
