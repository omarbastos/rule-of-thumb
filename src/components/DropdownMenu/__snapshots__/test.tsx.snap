// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownMenu /> should render DropdownMenu as default 1`] = `
<div>
  <div
    class="relative flex-none"
  >
    <button
      aria-label="View"
      class="w-40 bg-white text-sm py-2 px-4 whitespace-nowrap text-black border-solid border-2 border-black font-normal flex justify-center items-center"
      data-testid="button-dropdow-menu"
      type="button"
    >
      <span
        class="flex-auto"
      >
         
        Grid
      </span>
      <svg
        class=" flex-none justify-self-end"
        fill="none"
        height="7"
        viewBox="0 0 11 7"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.25 7L6.11959e-07 -9.17939e-07L10.5 0L5.25 7Z"
          fill="#333333"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
